Pass `browser_identity` on to privacy requests from consent request patches

allisonking opened the pull request:
Backend part of https://github.com/ethyca/fidesplus/issues/519

### Code Changes

* [x] Update the Identity model to support a new `user_id` 
* [x] Update the `providedidentitytype` enum to also support `user_id`
* [x] Update the patch endpoint to take a `browser_identity`
* [x] Update the `queue_privacy_request_to_propagate_consent` to merge identities 

### Steps to Confirm

* [ ] _list any manual steps for reviewers to confirm the changes_

### Pre-Merge Checklist

* [x] All CI Pipelines Succeeded
* Documentation:
  * [ ] documentation complete, [PR opened in fidesdocs](https://github.com/ethyca/fidesdocs/pulls)
  * [ ] documentation [issue created in fidesdocs](https://github.com/ethyca/fidesdocs/issues/new/choose)
* [ ] Issue Requirements are Met
* [ ] Relevant Follow-Up Issues Created
* [x] Update `CHANGELOG.md`

### Description Of Changes

_Write some things here about the changes and any potential caveats_

allisonking commented:
To verify the alembic migration, this query is useful:

```
select n.nspname as enum_schema,  
       t.typname as enum_name,  
       e.enumlabel as enum_value
from pg_type t 
   join pg_enum e on t.oid = e.enumtypid  
   join pg_catalog.pg_namespace n ON n.oid = t.typnamespace WHERE t.typname='providedidentitytype';
```

should yield 

```
 enum_schema |      enum_name       |  enum_value  
-------------+----------------------+--------------
 public      | providedidentitytype | email
 public      | providedidentitytype | phone_number
```
before migration, and after migration should have an additional `user_id` field. The downgrade should work too

pattisdr commented on src/fides/api/ctl/migrations/versions/392992c7733a_add_user_id_as_a_provided_identity_type.py:
```diff
@@ -0,0 +1,39 @@
+
+def upgrade():
+    with op.get_context().autocommit_block():
+        print("hi")
```
extra print statement

allisonking replied:
🤦‍♀️ at least it wasn't more embarrassing

pattisdr replied:
😄 I know what you mean

pattisdr commented on tests/ops/models/test_consent_request.py:
```diff
@@ -180,3 +181,50 @@ def test_do_not_queue_privacy_request_if_no_executable_preferences(
+        assert identity_of_privacy_request.email == "test@email.com"
+        assert identity_of_privacy_request.user_id == browser_identity.user_id
+
+        provided_identity.delete(mock_create_privacy_request)
```
This should be ` provided_identity.delete(db)` instead, I realize I made this same mistake above too.

pattisdr commented on src/fides/api/ops/api/v1/endpoints/consent_request_endpoints.py:
```diff
@@ -231,14 +231,16 @@ def queue_privacy_request_to_propagate_consent(
     """
-    identity = Identity()
+    # Create an identity based on any provided browser_identity
+    identity = browser_identity if browser_identity else Identity()
```
That's a good way to do this!

pattisdr commented on src/fides/api/ops/schemas/redis_cache.py:
```diff
@@ -11,6 +11,7 @@ class Identity(BaseSchema):
 
     phone_number: Optional[str] = None
     email: Optional[str] = None
+    user_id: Optional[str] = None
```
Have we confirmed this is the user_id that's needed here?  https://support.google.com/analytics/answer/3123669?hl=en, I just assumed that in the design docs, but I don't know for sure

pattisdr commented on tests/ops/models/test_consent_request.py:
```diff
@@ -180,3 +181,50 @@ def test_do_not_queue_privacy_request_if_no_executable_preferences(
+        assert mock_create_privacy_request.called
+        call_kwargs = mock_create_privacy_request.call_args[1]
+        identity_of_privacy_request = call_kwargs["data"][0].identity
+        assert identity_of_privacy_request.email == "test@email.com"
+        assert identity_of_privacy_request.user_id == browser_identity.user_id
```
This is a great test.  Another good test to add would be to the test consent request endpoints `test_set_consent_consent_preferences` and pass in a browser_identity dict and then assert that the user_id was added to identities retrieved in `consent_request.privacy_request.get_persisted_identity()`

pattisdr commented on src/fides/api/ctl/migrations/versions/392992c7733a_add_user_id_as_a_provided_identity_type.py:
```diff
@@ -0,0 +1,39 @@
+
+
+def upgrade():
+    with op.get_context().autocommit_block():
```
Are these needed specifically? From my understanding, they automatically commit all the migrations before this operation, potentially prematurely.

> Warning As is necessary, the database transaction preceding the block is unconditionally committed. This means that the run of migrations preceding the operation will be committed, before the overall migration operation is complete.
It is recommended that when an application includes migrations with “autocommit” blocks, that [EnvironmentContext.transaction_per_migration](https://alembic.sqlalchemy.org/en/latest/api/runtime.html#alembic.runtime.environment.EnvironmentContext.params.transaction_per_migration) be used so that the calling environment is tuned to expect short per-file migrations whether or not one of them has an autocommit block.

allisonking replied:
apparently the `ALTER TYPE` command [cannot be executed in a transaction](https://www.postgresql.org/docs/9.1/sql-altertype.html#AEN63173). I don't 100% understand it, but these answers go into some more detail: https://stackoverflow.com/questions/14845203/altering-an-enum-field-using-alembic

pattisdr replied:
That seems like it was for older postgres versions though

https://www.postgresql.org/docs/12/sql-altertype.html
> If ALTER TYPE ... ADD VALUE (the form that adds a new value to an enum type) is executed inside a transaction block, the new value cannot be used until after the transaction has been committed.

We do this in other migrations:
https://github.com/ethyca/fides/blob/main/src/fides/api/ctl/migrations/versions/906d7198df28_privacy_request_approve.py#L20

allisonking replied:
oh great, thanks for finding that!!

pattisdr reviewed (commented):
Bravo, thanks for knocking this out so quickly - just a few small changes requested

pattisdr commented on CHANGELOG.md:
```diff
@@ -38,6 +38,7 @@ The types of changes are:
 * Add the ability to execute Consent Requests via the Privacy Request Execution layer [#2125](https://github.com/ethyca/fides/pull/2125)
 * Privacy Center consent options can now be marked as `executable` in order to propagate consent requests [#2193](https://github.com/ethyca/fides/pull/2193)
 * Add a Mailchimp Transactional Consent Connector [#2194](https://github.com/ethyca/fides/pull/2194)
+* Add support for passing browser identities to consent request patches [#2304](https://github.com/ethyca/fides/pull/2304)
```
All these changelog items will need to be removed to unreleased at some point - but this can also wait until we merge `consent-propagation` next week too

pattisdr commented:
hey @allisonking just a heads-up that I updated the base `consent-propagation` branch with main after `unified-fides-resources` was merged because there were several conflicts there.  

I think the main consequences for you when getting up to date with `consent-propagation` are just the changelog and bumping the downrev of your migration.

allisonking commented:
Thanks, Dawn! I think I've addressed your comments now.

> I think the main consequences for you when getting up to date with consent-propagation are just the changelog and bumping the downrev of your migration.

I didn't see any alembic conflicts (didn't find another alembic file using that same downrev, and `alembic heads` only returned the one head) but let me know if I'm missing something!

pattisdr reviewed (approved):
> I didn't see any alembic conflicts (didn't find another alembic file using that same downrev, and alembic heads only returned the one head) but let me know if I'm missing something!

Whoops you're right, your migration is already pointed at `de456534dbda`, sorry!

This looks ready to merge into `consent-propagation`: thanks especially for the extra assertion in the consent request endpoints test.
